Align gRPC server span status codes to OTEL specs

FilipNikolovski opened the pull request:
# Description

This PR aligns the gRPC server span status codes, to be set based on the changes in the otel specification https://github.com/open-telemetry/opentelemetry-specification/pull/3333.

Fixes #1756

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Updated the `abort` test to check for proper span status against different grpc status codes.

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py:
```diff
@@ -158,25 +154,34 @@ def set_code(self, code):
+        error_status = Status(
+            status_code=StatusCode.ERROR, description=f"{code}:{details}"
+        )
+        status_codes = {
```
Thank you for putting in this PR that lines up with the [spec update](https://github.com/open-telemetry/opentelemetry-specification/pull/3333/files#diff-a4272fe55d12584e896526953d110354d9cc0c4ee26c6cdf559cbafdd50d9cea). The updated tests make sense.

tammy-baylis-swi commented:
Overall looks good. Please could you resolve the Changelog conflicts as there was a recent new release.

tammy-baylis-swi commented:
> Thanks! I've resolved the conflicts and the tests, and also updated the Aio grpc server to align the status codes properly with the specs.

Thank you @FilipNikolovski !

Please could you also address the lint suggestions: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019

I see there is also a celery test failure but I don't think it's related to changes in this PR: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019

FilipNikolovski commented:
> > Thanks! I've resolved the conflicts and the tests, and also updated the Aio grpc server to align the status codes properly with the specs.
> 
> Thank you @FilipNikolovski !
> 
> Please could you also address the lint suggestions: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019
> 
> I see there is also a celery test failure but I don't think it's related to changes in this PR: https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/8189053992/job/22392907101?pr=2019

Took me a while to resolve the failing tests & lint, but anyway..Cleaned up the code a little bit - moved the `_server_status` function to utilities, if you can please have another look :pray:.

lzchen reviewed (approved):
LGTM. It would be great to get this implemented for client as well :)

lzchen commented on CHANGELOG.md:
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
+### Fixed
+- Align gRPC span status codes to OTEL specification ([#1756](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1756))
```
Oops already merged but this should link to the PR :)

FilipNikolovski replied:
ah sorry, should I open up a new PR to fix the link?
